### ISSUE TITLE
Testers needed for config flow integration update

### ISSUE THREAD

**Issue by alexander0042** [alexander0042/pirate-weather-ha#57]
Hi everyone,

Since I launched this a year ago, I've been blown away by the enthusiasm and support I've received for this project! It's been amazing to hear all the different use cases, and the API has improved considerably over the past 12 months. 

One important update is to modernize this integration away from just relying on the Home Assistant Dark Sky integration to one that can be setup from the GUI, uses coordinated updates, and supports modern Home Assistant API calls. In short, this is that update. 

For anyone interested, I'm looking for some people to try out the new version and get some feedback on what works, what doesn't and where the rough edges are. From my testing, everything seems stable, but since this is a major change I want to test it out as much as possible before a wider release. 

As a starting point, this version should behave exactly the same way the production one currently does. If you have an entry in your configuration.yaml. then a new weather entity and/or sensors will be created, and should be identical. Anything else is a bug. Moreover, this integration should now be able to be set up from the GUI integrations page, producing the same integrations that currently exist. 

To install and test this version, you'll need to manually patch the "dev-config-flow" branch into your custom_components directory using SSH and restart Home Assistant Core:
```
cd "/config/custom_components/"
mv "/config/custom_components/pirateweather/manifest.json" "/config/custom_components/pirateweather/manifest.json.old"
mv "/config/custom_components/pirateweather" "/config/custom_components/pirateweather.old"
git clone --branch dev-config-flow https://github.com/alexander0042/pirate-weather-ha.git
mv "/config/custom_components/pirate-weather-ha/custom_components/pirateweather" "/config/custom_components/pirateweather"
```

The two outstanding known issues are cleaning up the text options for the monitored conditions, and updating the installation instructions in the README, but I'm sure more will crop up as people start using this version. 

Thanks again for the help, and let me know it goes!

**Comment by myopenflixr** [comment]
Hi @alexander0042. I'll be glad to help test.

I had to slightly modify your instructions based on my file structure and "think" I have the new version installed. I restarted Home Assistant and everything seems to be working as normal.

How can I verify that the new version is installed and working?

**Comment by alexander0042** [comment]
Thanks- and great question, should have included that in the initial post! Two signs that it's working:
1. You should see a warning in your log (after restarting HA core) that the yaml configurations have been imported and can be deleted
2. In the integrations page, there should be a new "Pirate Weather" box like this: 
![image](https://user-images.githubusercontent.com/26213938/196523823-c49be999-5a95-43f5-8cbe-4403c50a1412.png)

**Comment by myopenflixr** [comment]
OK. So when I "Check Configuration" in Developer Tools, I the following two errors:
`Platform error weather.pirateweather - Integration 'pirateweather' not found.`
`Platform error sensor.pirateweather - Integration 'pirateweather' not found.`

And when I go to integrations, I do not see the Pirate Weather Integration/Box. 

What should I do at this point?

**Comment by alexander0042** [comment]
Sorry this is trickier than it should be! This is after restarting HA Core (under "config/system")?

**Comment by myopenflixr** [comment]
Yes...And I can no longer restart HA Core due to the config error.
```
Failed to restart Home Assistant
The system cannot restart because the configuration is not valid: Platform error weather.pirateweather - Integration 'pirateweather' not found. Platform error sensor.pirateweather - Integration 'pirateweather' not found.
```

**Comment by myopenflixr** [comment]
OK. I did a hard CPU reboot and re-followed your instructions and all seems to be working.

So, according to logs, I can completely delete the pirateweather data from configuration.yaml now that it's been imported?

**Comment by alexander0042** [comment]
Again, sorry this was more complex than it needed to be! I'm (clearly) more comfortable with data wrangling than this sort of development. Happy this version worked!

And yes (although I'd just comment it out for now)! You can either leave the imported entities alone and they will continue to happily exist, or create a new one using the GUI at some point. One perk of this approach is you can now have as many Pirate Weather integrations as you'd like, so long as they're for different lat/longs, they'll have their own names and update coordinators.

**Comment by myopenflixr** [comment]
OK...I tried to configure the Integration to add TEMPERATURE as a "Monitored condition sensor" and got the following error:
`User input malformed: expected str for dictionary value @ data['forecast']`

Also, I'm not seeing a Humidity sensor within the list to add.

**Comment by alexander0042** [comment]
Ok then- this is exactly the sort of testing I was hoping for! I'll see where that humidity sensor went, and take a look at the options update logic

**Comment by myopenflixr** [comment]
OK....let me know when you have another update

I do rely on the temp and humidity sensors (nothing major, their just on my home page as entities)

**Comment by myopenflixr** [comment]
I was able to get Humidity added to the list of available sensors as well as correct the spelling mistake.

Spelling mistake fixed in `en.json - Line 40`
Humidity added to Available list of monitored conditions by adding the following line to `const.py` in the `ALL_CONDITIONS =,` section
`'humidity': 'humidity',`

However, I wouldn't know where to begin to fix the error: `User input malformed: expected str for dictionary value @ data['forecast']`

I'm sure you would have figured those two fixes out fairly quickly, but wanted to see if I could do it on my own :)

**Comment by alexander0042** [comment]
Good catch on the spelling mistake and the humidity, that's exactly the fix ! And don't sell yourself short- I love the challenge of trying to fix something without knowing what's going on, so always worth trying. 

The other issue is a lot trickier. It all comes down to Dark Sky having both a "weather" entity and a "sensor" entity, and the options flow trying to unload one when the other is what's in use. I think I have a fix for this, but want to look it over tomorrow morning, so I'll update when the branch is corrected. The net result will be that it'll take a delete and re-add to change from one to the other, but since most people are looking for both, that shouldn't be too much of an issue

**Comment by myopenflixr** [comment]
Your workaround worked. Weather is now working again.

I did go in and tried to add the w
Temperature and Humidity sensors. I got no errors when adding, but the sensors did not get added. I tried reloading the integration as well as restarting hass with no luck.

The integration still only showing 1 Service and 1 Entity.

**Comment by alexander0042** [comment]
Awesome, that's something at least! Next question is if your configuration.yaml is setup using a "sensor" block, "weather" block, or both? I'll need to clarify this in the instructions somehow, but if it's a "weather" block you can't add sensors (which is what shows up in monitored conditions), and visa versa. However, if it is configured as a sensor already, then that's an issue, since it should work

**Comment by myopenflixr** [comment]
Before your changes, I used a weather block and had all pirateweather config in weather.yaml. I have since commented out my weather block since hass pulled in pirateweather as an integration. I also commented out all pirateweather sensors used previously before this new configuration.

**Comment by alexander0042** [comment]
Ok, that makes sense- because you had both a weather block and a sensor block, you should have two separate integrations show up for Pirate Weather. If not, try un-commenting and restarting, which should bring them in. You should be able to add humidity to the integration for the sensors, which will have a several entities, compared to just one for the weather integration.

I do have an idea to simplify this somewhat and let sensors to added to a weather integration or visa versa, but not sure it's worth working into this particular update

**Comment by alexander0042** [comment]
Ok, think I've got this working now! The options flow is now a lot more robust, and can handle both platforms in one place! If you have time to try this one at some point, the best procedure is:

1. Delete any PW integrations in the UI
2. UN-comment the data blocks in the configuration.yaml
3. Follow the renaming and download steps in the original post 

Thanks again for helping here, it's invaluable to get this sort of feedback before a major change

**Comment by myopenflixr** [comment]
Followed instructions and now missing the main `weather.pirateweather` entity
It appears the only thing I have are all the pirateweather sensors

I'm also a little confused with what your trying to accomplish here. I thought the original goal was to get the pirateweather moved over to an integration thus eliminating any yaml configs.

But, let me know if you have any suggestions or fixes to bring back the `weather.pirateweather` entity as I'm completely missing my weather card on home screen.

**Comment by alexander0042** [comment]
I could definitely be clearer here. The goal is to move away from the yaml; however, by leaving the lines in it helps to test if the import is working, and deleting the existing UI integrations beforehand creates a "clean slate" for this testing. 

Did you notice anything in the logs when you restarted HA core? There should be something about importing the weather entity from the yaml, as well as an error of some sort if the import failed

**Comment by myopenflixr** [comment]
OK. That makes perfect sense.
Let me try again....I had since reverted back to the original.

**Comment by myopenflixr** [comment]
Here's what's in the Logs There's two separate messages.

```
This error originated from a custom integration.

Logger: custom_components.pirateweather.weather
Source: custom_components/pirateweather/weather.py:136
Integration: Pirate Weather (documentation, issues)
First occurred: 3:37:15 PM (1 occurrences)
Last logged: 3:37:15 PM

Configuration of Pirate Weather (weather entity) in YAML is deprecated Your existing configuration has been imported into the UI automatically and can be safely removed from your configuration.yaml file
```


```
This error originated from a custom integration.

Logger: custom_components.pirateweather.sensor
Source: custom_components/pirateweather/sensor.py:546
Integration: Pirate Weather (documentation, issues)
First occurred: 3:37:15 PM (1 occurrences)
Last logged: 3:37:15 PM

Configuration of Pirate Weather sensor in YAML is deprecated Your existing configuration has been imported into the UI automatically and can be safely removed from your configuration.yaml file
```

**Comment by alexander0042** [comment]
Ok, those are actually good things! They mean that the integration is reading both the sensor and weather blocks from the config file, so they should be importing. Downside is that they don't explain why the weather one isn't showing up in your setup.

If you create a new weather entity using the UI flow, does that one show up?

**Comment by myopenflixr** [comment]
How do I create a new weather entity using UI flow? Do you mean to try and add another Pirateweather "integration"?

**Comment by alexander0042** [comment]
Exactly that! Under the add integration button select Pirate Weather, and fill in the fields from there. To create a weather entity, you just need to make sure the "weather" checkbox is selected, and you can leave the rest blank

**Comment by myopenflixr** [comment]
Actually, I just looked at the integration already installed and it was missing the Weather checkbox....I checked that and got my main `weather.pirateweather` entity back!

Should I now try and comment out everything in yaml?

**Comment by myopenflixr** [comment]
I am surprised that was not checked by default.

**Comment by myopenflixr** [comment]
OK....So I commented out my weather block as well as all pirateweather sensors and restarted HASS. This resulted in the integration disappearing. I then went to integrations and manually added the pirateweather integration. I checked weather and sensors and selected the temperature & humidity sensors I needed and it all added everything as expected.

I restarted HASS and everything remained. It appears that all is working as expected.

**Comment by myopenflixr** [comment]
I do have a feature request.

I would like to have the ability to not show anything after decimal point for temperature and humidity. I had to create `value_templates` to get my temperature and humidity to display whole numbers (rounded). I would love to have this feature within pirateweather so I can get rid of my `value_templates`

**Comment by coaster3000** [comment]
While I am testing this, would you like any  issue reports on this thread, or would you like an issue created referencing this issue?

**Comment by alexander0042** [comment]
Great feedback, this is exactly the sort of stuff I was looking for!

1. Two instances is a really good point- one of the goals for this change was to make it easier to create multiple instances, so it's definitely important! Currently, the integration checks to see if there's already a configured instance with the same location and doesn't create another entity if there already is one in place, which works well enough, but I'll add a check for daily/ hourly modes. 

2. I think this is a documentation/ working fix. If you're just creating a weather entity, then there should only be a weather entity. If someone is looking to create sensor entities (which was something supported by the original Dark Sky), then it'll create a sensor for every requested variable at every requested time (which could be a lot). My plan is to clarify this in the setup UI flow with something along the lines of "Don't select sensor unless you want individual entities for the following options"

3. That's a known bug that I'm still working on. You can manually delete these orphaned entities without issue, and at least they're clearly marked, but I'm going to have to ask in the discord about this one, since I've tried every way I can think of and nothing is working to address this.

Hoping to have a new version Monday/ Tuesday next week that adds a rounding option (great suggestion from @myopenflixr!) as well as these two changes

**Comment by coaster3000** [comment]
@alexander0042 for number 2, I understand what you mean there. However, that is not what I meant by it creating a large quantity of devices. Homeassistant lists devices and entities. Devices can have entities associated with it, and what I see happening is, for every entity created when using the entities option to track different stuff, each entity will have it's own device.

Now I don't know if its a limitation in design, but what I would expect is for all entities under the same integration config / instance, those entities would be grouped under a single device. If this isn't possible due to implementation, I would suggest giving the devices names that correspond to the entity as its hard hunting for the exact device responsible for the entity.

**Comment by coaster3000** [comment]
I haven't been able to dig deep into what changed in the code yet but for some reason all of my wind Gust readings are incorrect. I have my units in the gui set to us but the wind gust never goes above 1.0. it's always below 1 which has made no sense.

If I find the change, I'll reference the line for you.

**Comment by alexander0042** [comment]
Ok, new version is out! Changes in: https://github.com/alexander0042/pirate-weather-ha/commit/9287401dd986fac45fa990c8baf0841b4fc1b10a 

- Added rounding (credit to @myopenflixr)
- Multiple weather entities (@coaster3000)
- Removed the debug warning (@coaster3000)
- Fixed the wind speeds (unit conversion error) (@coaster3000)
- Cleaned up some of the language in the config flow prompt

I checked how the existing integration does it, and since it creates separate devices for each time/ condition, I think it's best to leave it for compatibility. The "one device many entities" approach is definitely the way things are going, so I'm hoping that once this is released and people are migrated over, I can switch it to a single device then!

**To update to this version:**
1. Delete any existing Pirate Weather Integrations using the UI (one time step since the config properties have changed)
2. Uncomment any yaml entries you'd like imported again
3. Run the update steps:
```
cd "/config/custom_components/"
mv "/config/custom_components/pirateweather/manifest.json" "/config/custom_components/pirateweather/manifest.json.old"
mv "/config/custom_components/pirateweather" "/config/custom_components/pirateweather.old"
git clone --branch dev-config-flow https://github.com/alexander0042/pirate-weather-ha.git
mv "/config/custom_components/pirate-weather-ha/custom_components/pirateweather" "/config/custom_components/pirateweather"
```

**Comment by coaster3000** [comment]
_I apologize for any grammar or odd phrasing, as I am dictating this reply on my phone since I am not at a computer at the moment_

A successful update has been performed. I differ from the instructions a little bit I like to keep all back up so I just renamed my old directory to old2.

I uncommented my original yaml configuration, which contains two weather integrations, one for hourly, one for daily. I also uncommented my sensor configuration.

TLDR:
- can only create through GUI one integration
- does not properly import config files that include more than one integration, including sensors.
- only the last configuration will import.

- _oddly when more than two integrations are present, two will be imported (my config includes three which includes the sensor section)_

As for the results of the import.

- my configuration has two weather integrations one named home and one named home hourly.
Home is listed first hourly is listed second in the config. When imported only the hourly config is imported properly, the daily one is not imported at all.

My configuration also has a section for sensors, which is imported alongside the hourly integration.

- when attempting to comment out the second weather integration, and redo the import after deleting the integrations. It will result in only the sensors being imported, if the only two integrations that are not commented are the sensors and a weather integration. 

- deleting the integration and commenting out the whole config, then proceeding to reboot and then create the integration yourself. It will allow you to create only one integration and you could check both the sensors and weather options. If you attempt to create a second integration it will allow you to enter the configuration details but once you continue it will give you an error saying the location is already configured. (Paraphrased that, I am dictating my report a the moment since I'm not at a computer.)

**Comment by alexander0042** [comment]
Ok, this is super helpful feedback! Do you think you could paste the relevant sections of your configuration.yaml (without the API key) here so I could try it on my end? That would really help to debug this, since the issues you're hitting here are pretty important!

**Comment by coaster3000** [comment]
Here is the requested config with api and location info stripped.
```yaml
weather:
  - platform: pirateweather
    api_key: ""
    # Additional optional values:
#    latitude: 
#    longitude: 
    #mode: "hourly"  (default) or "daily"
    mode: "daily"
    name: Home
  - platform: pirateweather
    api_key: ""
    # Additional optional values:
#    latitude: 
#    longitude: 
    #mode: "hourly"  (default) or "daily"
    mode: "hourly"
    name: Home Hourly
    

# you can also get a sensor data
sensor:
  - platform: pirateweather
    api_key: ""
    scan_interval: '00:15:00'
    monitored_conditions:
      - alerts
      - nearest_storm_distance
      - temperature
      - precip_probability
      - precip_type
      - humidity
      - cloud_cover
      - precip_intensity
      - wind_speed
      - wind_gust
```

**Comment by coaster3000** [comment]
Hey, quick update. I found a possible error in my configuration that may have been producing some inconsistencies in the sensor portion of the config. I didn't notice the error at all because it isn't something that home assistant considers an error; however, I could see it producing issues.

It was a duplicate key within the configuration.yaml for ```sensor: ``` , the first one directs homeassistant to include the sensors.yaml file while the second was the pirateweather platform.

I will get back to you on the results of this as its pretty late and I'm not up for retesting the import at the moment.

**Comment by alexander0042** [comment]
Oh, that's interesting- it's an easy mistake to make, I also had duplicate sensor blocks in mine at one point. Thanks for catching it, and let me know what happens once that's resolved

**Comment by coaster3000** [comment]
The removal ofthe duplicate key in config has no effect on the bug. The bug is still present.

**Comment by mattlward** [comment]
I have installed the latest version of the integration.  I have an issue, I am running on a test system and only had the weather block in the configuration.yaml file when I installed.  It loaded and worked.  But, since never having this installed before, I attempted to check both Sensor and Weather and assigned a few sensors and rebooted.  Still have a good forecast but the sensors I added never appeared.

![image](https://user-images.githubusercontent.com/8540074/199852611-59c2de4d-3fff-4d41-ae58-2d3f9297b706.png)

So, as a followup attempt I added the following to my sensors file and removed the sensors from the UI and rebooted expecting the yaml sensors to be imported.  Did not gain new sensors upon reboot.

Added sensor config:
```
- platform: pirateweather
  api_key: "9snAKslklf4cKDXfZYGDc6sC3zhAVs6rBjuBi7Qg"
  scan_interval: '00:15:00'
  monitored_conditions:
    - alerts
    - nearest_storm_distance
    - temperature
    - precip_probability
    - precip_type
    - humidity
    - cloud_cover
    - precip_intensity
    - wind_speed
    - wind_gust
 ```

**Comment by mattlward** [comment]
Is the source of data the same as DarkSky?  I have noticed that my default location data from DarkSky and PirateWeather differ a few degrees.  And I am not sure how to kill off DarkSky, I have removed it from the weather and sensor blocks but it keeps updating.

**Comment by mattlward** [comment]
The sensor key "    - apparent_temperature" appears to not be considered valid in my initial sensor block...

```
Invalid config for [sensor.pirateweather]: apparent_temperature is not a valid option for dictionary value @ data['monitored_conditions']. Got ['alerts', 'nearest_storm_distance', 'temperature', 'precip_probability', 'precip_type', 'humidity', 'cloud_cover', 'precip_intensity', 'wind_speed', 'wind_gust', 'apparent_temperature']. (See ?, line ?). 
```

**Comment by alexander0042** [comment]
Thanks @mattlward, thanks for moving the thread over here- helps me to keep everything in one place, and since I have a 2-month old, I need all the reminders I can get! 

I'll take a look at the apparently temperature sensor today and see if I can solve that one! As far as a data source goes, that's the coolest part of this integration (by far). This is a whole separate weather forecast API (which I'm way more comfortable with than this sort of coding), which creates a forecast directly from NOAA models. So it should be similar to Dark Sky, but likely won't ever be the same

**Comment by alexander0042** [comment]
Thank you again to everyone who has been testing this! To respond to questions:
 - @mattlward, that's a good catch. I'll clarify the documents to explain that if multiple weather entities are needed, they need to have different names, This shouldn't be too much of a limitation, since it would get way too complicated with things having the same name anyway. 
 - @mattlward, another good catch with the apparent temperature issue! This one was just a typo, and should be fixed now!
 - @coaster3000, this one was interesting! The root cause stemmed from you doing something I didn't think was possible- having a sensor with no hourly or daily forecast times configured. This seems to work, but since I didn't realize it was possible, I used these as a shorthand for checking if something was a sensor or weather entity. I've added a name check to the unique ID, which lets the setup know that there's something new to add!
 - @myopenflixr, rounding should be improved now! I switched it to return an integer for the rounded value, as opposed to just rounding it, which should remove the decimal. Just to clarify, you were talking about the sensors here, not the weather entity?
 
 New version will be pushed shortly! **Please DELETE** any existing integrations and uncomment your configuration.yaml before following the steps. Because the unique-id has changed, it needs to start this from scratch.

**Comment by alexander0042** [comment]
Just passing along another update here- I pushed a new change that cleans up some of the language, clarified some warnings, and fixes the scan interval parameter, so two more items off the list!

I'm considering if it's worth keeping the options configuration around? It would be nice to be able to change monitored conditions after the initial setup; however, since I can't seem to correct the orphan entity issue, I'm worried that it might create more confusion than it's worth compared to just deleting and recreating an integration. I sent a message to discord, so we'll see if they have any suggestions. The other idea would be to keep the options flow, but only allow the name, location, forecast mode, units, and rounding to be changed? This would avoid any entity deletion issues, while allowing a few things to be edited.

**Comment by wyliletv** [comment]
HI. Thanks for this! I am running this development version now. I really like the simple way of checking sensors to add them.
I'm new to HA and a long time Samsung SmartThings user.
I got confused in the instructions with which Integration page -the  HACS Integrations or the main Integrations. I'm still learning the nomenclature in HA, so that has a factor :)

I experimented with the 'replacement for DarkSky' idea a while, thinking it is something I NEEDED to do for HA to work well. Is that necessary if I started using HA in October 2022? I have come to the conclusion I do not.

Thanks again!!

**Comment by coaster3000** [comment]
Finally updated the dev version, I have been pretty busy lately.

The import ran exactly as expected this time around. I will fiddle around and see if there is anything left I can find. 

One thing I found so far with the config flow setup is the ability to create an integration without specifying weather or sensors to be used. Understandably, this is user error; however, it may be beneficial to warn the user about this mistake during the setup process if that is possible.

It will reduce unwanted complaints / bug reports on the matter.

**Comment by alexander0042** [comment]
That's a really good point about the flow setup check boxes, and I can foresee issues stemming from that. I'll change it so at least one has to be selected, since it really wouldn't make sense otherwise

**Comment by alexander0042** [comment]
And we're live! I pushed this update out via HACS today, so I'm going to close this thread for now. If (when) other issues come up with the update, please open new issues and we can tackle them one by one.

Thank you again for everyone for helping to debug this, I really appreciate the help, and wouldn't have made it to a release without the feedback here,
